src/rewritable_units/text_encoder.rs

```diff
@@ -0,0 +1,183 @@
+            non_utf8_encoder: (encoding != UTF_8).then(|| TextEncoder::new(encoding)),
+            output_handler,
+        }
+    }
+
+    /// Writes the given UTF-8 string to the output, converting the encoding and [escaping](ContentType) if necessary.
+    ///
+    /// It may be called multiple times. The strings will be concatenated together.
+    #[inline]
+    pub fn write_str(&mut self, content: &str, content_type: ContentType) {
+        match content_type {
+            ContentType::Html => self.write_html(content),
+            ContentType::Text => self.write_body_text(content),
+        }
+    }
+
+    pub(crate) fn write_html(&mut self, html: &str) {
+        if let Some(encoder) = &mut self.non_utf8_encoder {
+            encoder.encode(html, self.output_handler);
+        } else if !html.is_empty() {
+            (self.output_handler)(html.as_bytes());
+        }
+    }
+
+    /// For text content, not attributes
+    pub(crate) fn write_body_text(&mut self, plaintext: &str) {
+        if let Some(encoder) = &mut self.non_utf8_encoder {
+            escape_body_text(plaintext, &mut |chunk| {
+                debug_assert!(!chunk.is_empty());
+                encoder.encode(chunk, self.output_handler);
+            });
+        } else {
+            escape_body_text(plaintext, &mut |chunk| {
+                debug_assert!(!chunk.is_empty());
+                (self.output_handler)(chunk.as_bytes());
+            });
+        }
+    }
+
+    #[inline]
+    pub(crate) fn output_handler(&mut self) -> &mut dyn FnMut(&[u8]) {
+        &mut self.output_handler
+    }
+}
+
+enum Buffer {
+    Heap(Vec<u8>),
+    Stack([u8; 63]), // leave a byte for the tag
+}
+
+struct TextEncoder {
+    encoder: Encoder,
+    buffer: Buffer,
+}
+
+impl TextEncoder {
+    #[inline]
+    pub fn new(encoding: &'static Encoding) -> Self {
+        debug_assert!(encoding != UTF_8);
+        debug_assert!(encoding.is_ascii_compatible());
+        Self {
+            encoder: encoding.new_encoder(),
+            buffer: Buffer::Stack([0; 63]),
+        }
+    }
+
+    /// This is more efficient than `Bytes::from_str`, because it can output non-UTF-8/non-ASCII encodings
+    /// without heap allocations.
+    /// It also avoids methods that have UB: https://github.com/hsivonen/encoding_rs/issues/79
+    #[inline(never)]
+    fn encode(&mut self, mut content: &str, output_handler: &mut dyn FnMut(&[u8])) {
+        loop {
+            debug_assert!(!self.encoder.has_pending_state()); // ASCII-compatible encodings are not supposed to have it
+            let ascii_len = Encoding::ascii_valid_up_to(content.as_bytes());
+            if let Some((ascii, remainder)) = content.split_at_checked(ascii_len) {
+                if !ascii.is_empty() {
+                    (output_handler)(ascii.as_bytes());
+                }
+                if remainder.is_empty() {
+                    return;
+                }
+                content = remainder;
+            }
```

Comment:
Let's refactor this part in a separate function with a meningful name that reflects that it's a fast path fro leading ASCII bytes